Diagnostics throws NPE for invalid code

**Description:**
Diagnostics listener throws NPE for the following invalid code;
```ballerina
type Person record {|
   string firstName;
   string lastName;
   int age;
|};

public function main() {
    string g;
    string h;
    int i;
    {firstName:g, lastName:h, _} = fooPerson(1, false);
}

function fooPerson(float a, boolean b) returns Person {
    return {firstName:"john", lastName:"williams", age:40};
}

```
**Error:**

```
Caused by: java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLog.reportDiagnostic(BLangDiagnosticLog.java:117)
	at org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLog.error(BLangDiagnosticLog.java:87)
	at org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLogHelper.error(BLangDiagnosticLogHelper.java:65)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.checkRecordVarRefEquivalency(SemanticAnalyzer.java:1829)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1744)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangRecordDestructure.accept(BLangRecordDestructure.java:65)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2610)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2578)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:2574)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:361)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2610)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:333)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2610)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2578)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:2570)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:256)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:161)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:230)
	at org.wso2.ballerinalang.compiler.CompilerDriver.typeCheck(CompilerDriver.java:322)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:256)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:248)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:140)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:166)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:174)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:91)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:79)
	at org.ballerinalang.langserver.compiler.LSModuleCompiler.compileSafe(LSModuleCompiler.java:232)
	... 9 more

[Error - 4:20:50 PM] Computing 'diagnostics' failed! {uri: '/Users/rasika/Documents/WSO2/Ballerina/samples/parser-samples/hello-world/main.bal', error: 'Oh no, something really went wrong. Bad. Sad.'}
org.ballerinalang.langserver.compiler.exception.CompilationFailedException: Oh no, something really went wrong. Bad. Sad.
	at org.ballerinalang.langserver.compiler.LSModuleCompiler.compileSafe(LSModuleCompiler.java:246)
	at org.ballerinalang.langserver.compiler.LSModuleCompiler.getBLangPackages(LSModuleCompiler.java:177)
	at org.ballerinalang.langserver.diagnostic.DiagnosticsHelper.compileAndSendDiagnostics(DiagnosticsHelper.java:76)
	at org.ballerinalang.langserver.BallerinaTextDocumentService.lambda$didChange$16(BallerinaTextDocumentService.java:747)
	at org.ballerinalang.langserver.util.Debouncer.lambda$call$0(Debouncer.java:42)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLog.reportDiagnostic(BLangDiagnosticLog.java:117)
	at org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLog.error(BLangDiagnosticLog.java:87)
	at org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLogHelper.error(BLangDiagnosticLogHelper.java:65)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.checkRecordVarRefEquivalency(SemanticAnalyzer.java:1829)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1744)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangRecordDestructure.accept(BLangRecordDestructure.java:65)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2610)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2578)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:2574)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:361)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2610)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:333)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2610)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2578)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:2570)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:256)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:161)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:230)
	at org.wso2.ballerinalang.compiler.CompilerDriver.typeCheck(CompilerDriver.java:322)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:256)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:248)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:140)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:166)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:174)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:91)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:79)
```

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**
<!-- Any related issues such as sub tasks, issues reported in other repositories (e.g component repositories), similar problems, etc. -->

**Suggested Labels (optional):**
<!-- Optional comma separated list of suggested labels. Non committers can’t assign labels to issues, so this will help issue creators who are not a committer to suggest possible labels-->

**Suggested Assignees (optional):**
<!--Optional comma separated list of suggested team members who should attend the issue. Non committers can’t assign issues to assignees, so this will help issue creators who are not a committer to suggest possible assignees-->

--- a/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java
+++ b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java
@@ -2532,6 +2532,7 @@ public class BLangNodeTransformer extends NodeTransformer<BLangNode> {
                 (BLangRecordDestructure) TreeBuilder.createRecordDestructureStatementNode();
         recordDestructure.varRef = (BLangRecordVarRef) createExpression(assignmentStmtNode.varRef());
         recordDestructure.setExpression(createExpression(assignmentStmtNode.expression()));
+        recordDestructure.pos = getPosition(assignmentStmtNode);
         return recordDestructure;
     }
 

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java
@@ -61,4 +61,13 @@ public class RecordDefNegativeTest {
         BAssertUtil.validateError(compileResult, indx++, "incompatible types: expected 'string', found 'int'", 79, 16);
         assertEquals(compileResult.getErrorCount(), indx);
     }
+
+    @Test(description = "Test record destructure negative cases")
+    public void recordDestructureTest() {
+        CompileResult compileResult = BCompileUtil.
+                compile("test-src/record/negative/record-destructure-negative.bal");
+        BAssertUtil.validateError(compileResult, 0,
+                "invalid record binding pattern; unknown field '_' in record type 'Person'",
+                27, 5);
+    }
 }

--- a/tests/jballerina-unit-test/src/test/resources/test-src/record/negative/record-destructure-negative.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/record/negative/record-destructure-negative.bal
@@ -1,0 +1,32 @@
+// Copyright (c) 2021 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+type Person record {|
+   string firstName;
+   string lastName;
+   int age;
+|};
+
+public function main() {
+    string g;
+    string h;
+    int i;
+    {firstName:g, lastName:h, _} = fooPerson(1, false);
+}
+
+function fooPerson(float a, boolean b) returns Person {
+    return {firstName:"john", lastName:"williams", age:40};
+}

